storage/storage_interface.ml

```diff
@@ -100,441 +186,1076 @@ let parse_nbd_uri nbd =
-	type id =
-		| Task of Task.id
-		| Vdi of vdi
-		| Dp of dp
-		| Mirror of Mirror.id
-
-	type t =
-		| Task_t of Task.id * Task.t
-		| Vdi_t of vdi * vdi_info
-		| Dp_t of dp * dp_stat_t
-		| Mirror_t of Mirror.id * Mirror.t
+  type id =
+    | Task of Task.id
+    | Vdi of Vdi.t
+    | Dp of dp
+    | Mirror of Mirror.id
+  [@@deriving rpcty]
+
+  type t =
+    | Task_t of Task.id * Task.t
+    | Vdi_t of Vdi.t * vdi_info
+    | Dp_t of dp * dp_stat_t
+    | Mirror_t of Mirror.id * Mirror.t
+  [@@deriving rpcty]
+
+  let rpc_of_id = Rpcmarshal.marshal id.Rpc.Types.ty
+end
 
+type uuid = string [@@deriving rpcty]
+
+module Errors = struct
+  type error =
+    | Backend_error_with_backtrace of (string * (string list)) (** name * params *)
+    | Sr_not_attached of string               (** error: SR must be attached to access VDIs *)
+    | Vdi_does_not_exist of string            (** error: the VDI is unknown *)
+    | Illegal_transition of (Vdi_automaton.state * Vdi_automaton.state) (** This operation implies an illegal state transition *)
+    | Backend_error of (string * (string list)) (** error: of the form SR_BACKEND_FAILURE *)
+    | Does_not_exist of (string * string)
+    | Cancelled of string
+    | Redirect of string option
+    | Sr_attached of string
+    | Unimplemented of string
+    | Activated_on_another_host of uuid
+    | Duplicated_key of string
+    | No_storage_plugin_for_sr of string
+    | Content_ids_do_not_match of (string * string)
+    | Missing_configuration_parameter of string
+    | Internal_error of string
+    | Unknown_error [@@default Unknown_error] [@@deriving rpcty]
 end
 
-type uuid = string
-
-exception Backend_error_with_backtrace of (string * (string list)) (** name * params *)
-
-exception Sr_not_attached of string               (** error: SR must be attached to access VDIs *)
-exception Vdi_does_not_exist of string            (** error: the VDI is unknown *)
-exception Illegal_transition of (Vdi_automaton.state * Vdi_automaton.state) (** This operation implies an illegal state transition *)
-exception Backend_error of (string * (string list)) (** error: of the form SR_BACKEND_FAILURE *)
-exception Does_not_exist of (string * string)
-exception Cancelled of string
-exception Redirect of string option
-exception Sr_attached of string
-exception Unimplemented of string
-exception Activated_on_another_host of uuid
-exception Duplicated_key of string
-exception No_storage_plugin_for_sr of string
-exception Content_ids_do_not_match of (string * string)
-exception Missing_configuration_parameter of string
+exception Storage_error of Errors.error
+
+let () =
+  (* register printer *)
+  let sprintf = Printf.sprintf in
+  let string_of_error e =
+    Rpcmarshal.marshal Errors.error.Rpc.Types.ty e |> Rpc.to_string
+  in
+  let printer = function
+    | Storage_error e ->
+      Some
+        (sprintf "Storage_error (%s)" (string_of_error e))
+    | _ -> None
+  in
+  Printexc.register_printer printer
```

Comment:
Should we factor this out into a helper used by the other ppx-ed interfaces? (well I guess its a little late now to be actually useful, since this is the last ppx-ed one).

Reply:
Yes, into ocaml-rpc along with the type combinators.